src/data/transforms/normalize.rs

```diff
@@ -0,0 +1,158 @@
+//! The Normalizing Transformer
+//!
+//! This module contains the `Normalizer` transformer.
+//!
+//! The `Normalizer` transformer is used to transform input data
+//! so that the l2 norm of each rows are all equal to 1.
+//! If input data has a row with all 0, `Normalizer` keeps the row as it is.
+//!
+//! Because transformation is performed per row independently,
+//! inverse transformation is not supported.
+//!
+//! # Examples
+//!
+//! ```
+//! use rusty_machine::data::transforms::{Transformer, Normalizer};
+//! use rusty_machine::linalg::Matrix;
+//!
+//! // Constructs a new `Normalizer`
+//! let mut transformer = Normalizer::default();
+//!
+//! let inputs = Matrix::new(2, 2, vec![-1.0, 2.0, 1.5, 3.0]);
+//!
+//! // Transform the inputs
+//! let transformed = transformer.transform(inputs).unwrap();
+//! ```
+
+use learning::error::{Error, ErrorKind};
+use linalg::{Matrix, BaseMatrix, BaseMatrixMut};
+use super::Transformer;
+
+use libnum::Float;
+
+/// The Normalizer
+///
+/// The Normalizer provides an implementation of `Transformer`
+/// which allows us to transform the all rows to have the same l2 norm (1).
+///
+/// See the module description for more information.
+#[derive(Debug)]
+pub struct Normalizer;
+
+/// Create a `Normalizer`.
+impl Default for Normalizer {
+    fn default() -> Self {
+        Normalizer {}
+    }
+}
+
+impl Normalizer {
+    /// Constructs a new `Normalizer`, `Normalizer` takes no parameters for initialization.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use rusty_machine::data::transforms::Normalizer;
+    ///
+    /// // Constructs a new `Normalizer`
+    /// let _ = Normalizer::new();
+    /// ```
+    pub fn new() -> Self {
+        Normalizer::default()
+    }
+}
+
+impl<T: Float> Transformer<Matrix<T>> for Normalizer {
+    fn transform(&mut self, mut inputs: Matrix<T>) -> Result<Matrix<T>, Error> {
+        let dists: Vec<T> = inputs.iter_rows().map(|x| dist(x)).collect();
```

Comment:
I believe we can do this more efficiently than the `dist` function. Vectorizing the dot product takes some work, which is why we have the [`utils::dot`](https://athemathmo.github.io/rulinalg/doc/rulinalg/utils/fn.dot.html) function in rulinalg.

Once we merge rulinalg 0.4.x we can update to using the `MatrixNorm` trait directly here.

Oh, and just in case I am wrong and the `dist` function is better. This line can be written as:

```rust
let dists: Vec<T> = inputs.iter_rows().map(dist).collect();
```